### PR TITLE
support logind's lock signals

### PR DESCRIPTION
**Pull request by boucman**

### CODE REVIEW

**Comment by emersion** [comment]
Can you rebase this PR against master?

File: main.c [hunk 1]
```diff
@@ -205,27 +230,86 @@ static int dbus_event(int fd, uint32_t mask, void *data) {
...
+		swayidle_log_errno(LOG_ERROR, "Failed to add D-Bus signal match : sleep");
 		return;
 	}
 	acquire_sleep_lock();
+}
 
-	struct wl_event_source *source = wl_event_loop_add_fd(state.event_loop,
-		sd_bus_get_fd(bus), WL_EVENT_READABLE, dbus_event, bus);
-	wl_event_source_check(source);
+static void setup_lock_listener(void) {
```
Comment (emersion) [inline main.c]: There's some overlap between these functions, could probably share some code

File: main.c [hunk 2]
```diff
@@ -178,6 +180,29 @@ static int prepare_for_sleep(sd_bus_message *msg, void *userdata,
 
 	return 0;
 }
+static int prepare_for_lock(sd_bus_message *msg, void *userdata,
```
Comment (emersion) [inline main.c]: Should probably renamed those to `handle_lock`

File: main.c [hunk 3]
```diff
@@ -371,6 +465,46 @@ static int parse_resume(int argc, char **argv) {
...
+}
+
+static int parse_unlock(int argc, char **argv) {
+#if !HAVE_SYSTEMD && !HAVE_ELOGIND
+	swayidle_log(LOG_ERROR, "unlock not supported: swayidle was compiled"
+			" with neither systemd nor elogind support.");
+	exit(-1);
+#endif
+	if (argc < 2) {
+		swayidle_log(LOG_ERROR, "Too few parameters to after-resume command. "
```
Comment (emersion) [inline main.c]: s/after-resume/unlock/?

File: main.c (hunk 2 above)
Comment (boucman) [reply main.c]: I'll rename all three while i'm at it

File: main.c (hunk 1 above)
Comment (boucman) [reply main.c]: you mean setup_lock_listener, setup_unlock_listener ?

yeah, a bit... but I would be suprised if there were a third user, which would make sharing worth it...
i'll see what I can do anyway, but It won't share more than a couple of lines

File: main.c (hunk 3 above)
Comment (boucman) [reply main.c]: urgh, my bad

File: main.c (hunk 2 above)
Comment (emersion) [reply main.c]: Well `prepare_for_sleep` really handles the `PrepareForSleep` event, so it'd make sense to keep it as is.

File: main.c (hunk 1 above)
Comment (emersion) [reply main.c]: Could share the `GetSessionByPID` stuff

**Comment by boucman** [comment]
OK. V2 fixes what you asked.
I moved the session detection to  the connect_to_dbus function, to avoid having a function just for that.

File: main.c [hunk 4]
```diff
@@ -205,27 +231,74 @@ static int dbus_event(int fd, uint32_t mask, void *data) {
...
+	}
 
-	ret = sd_bus_match_signal(bus, NULL, "org.freedesktop.login1",
+	ret = sd_bus_message_read(msg, "o", &session_name_tmp);
+	if (ret < 0) {
+		swayidle_log(LOG_ERROR,
+				"Failed to read session name\n");
+		goto cleanup;
+	}
+	session_name= strdup(session_name_tmp);
```
Comment (emersion) [inline main.c]: Style: missing space before `=`

Comment (boucman) [reply main.c]: fixed

**Comment by emersion** [comment]
Thanks!

**Pull request merged**